### ISSUE TITLE
Frontend API design

### ISSUE THREAD

**Issue by merrymercy** [alpa-projects/alpa#1]
- Should we know the parallelization strategy before weight initialization?

| | Know | Do now know|
| --- | --- | --- |
| Weight reorganization overhead | Zero | High |
| Programming Style | Declarative | Imperative |

If we know the parallelization strategy before weight initialization, we can put the weights into the right places at the first time. Otherwise, we have to move tensors after we choose the parallelization strategy.
However, if we want to know the parallelization strategy before weight initialization, we have to see the whole training graph before initialization. This is not an intuitive and imperative programming style.

- Should we know which tensor gradients/weights are? 

| | Know | Do now know|
| --- | --- | --- |
| Difficulty to implement some specific strategies | Easy | Hard |
| Generalization | Low | High |

If we know which tensors gradients are, we can easily implement strategies like data-parallel and insert correct allreduce.
If we do not make the assumption that they are gradients, but just treat them as valina intermediate results, we have to do a more general analysis.

- Should we see the training loop?
If we can see the training loop, then we know the weight tensors are updated iteratively, we can make their update in place and preserves their layouts. However, if we want to see the training loop, we have to compile all complicated logic like data loading, preprocessing, model checkpointing, and logging into our IR, which is impossible.
